polyfill/lib/calendar.mjs

```diff
@@ -333,53 +332,1354 @@ function resolveNonLunisolarMonth(calendarDate) {
+    let addedCalendarDate;
+    do {
+      addedIsoDate = this.addDaysIso(addedIsoDate, increment);
+      addedCalendarDate = this.isoToCalendarDate(addedIsoDate, cache);
+    } while (addedCalendarDate.month === originalCalendarMonth);
+
+    // Now back up to the last day of the original month
+    const endOfMonthIso = this.addDaysIso(addedIsoDate, -addedCalendarDate.day);
+    const endOfMonthCalendar = this.isoToCalendarDate(endOfMonthIso, cache);
+    return endOfMonthCalendar.day;
+  },
+  daysInPreviousMonth(calendarDate, cache) {
+    const { day, month, year } = calendarDate;
+
+    // Check to see if we already know the month length, and return it if so
+    const previousMonthYear = month > 1 ? year : year - 1;
+    let previousMonthDate = { year: previousMonthYear, month, day: 1 };
+    const previousMonth = month > 1 ? month - 1 : this.monthsInYear(previousMonthDate, cache);
+    previousMonthDate = { ...previousMonthDate, month: previousMonth };
+    const min = this.minimumMonthLength(previousMonthDate);
+    const max = this.maximumMonthLength(previousMonthDate);
+    if (min === max) return max;
+
+    const isoDate = this.calendarToIsoDate(calendarDate, 'constrain', cache);
+    const lastDayOfPreviousMonthIso = this.addDaysIso(isoDate, -day);
+    const lastDayOfPreviousMonthCalendar = this.isoToCalendarDate(lastDayOfPreviousMonthIso, cache);
+    return lastDayOfPreviousMonthCalendar.day;
+  },
+  startOfCalendarYear(calendarDate) {
+    return { year: calendarDate.year, month: 1, day: 1 };
+  },
+  startOfCalendarMonth(calendarDate) {
+    return { year: calendarDate.year, month: calendarDate.month, day: 1 };
+  },
+  calendarDaysUntil(calendarOne, calendarTwo, cache) {
+    const oneIso = this.calendarToIsoDate(calendarOne, 'constrain', cache);
+    const twoIso = this.calendarToIsoDate(calendarTwo, 'constrain', cache);
+    return this.isoDaysUntil(oneIso, twoIso);
+  },
+  isoDaysUntil(oneIso, twoIso) {
+    const duration = ES.DifferenceDate(
+      oneIso.year,
+      oneIso.month,
+      oneIso.day,
+      twoIso.year,
+      twoIso.month,
+      twoIso.day,
+      'days'
+    );
+    return duration.days;
+  },
+  // The short era format works for all calendars except Japanese, which will
+  // override.
+  eraLength: 'short',
+  // All built-in calendars except Chinese/Dangi and Hebrew use an era
+  hasEra: true,
+  monthDayFromFields(fields, overflow, cache) {
+    let { year, month, monthCode, day, era, eraYear } = fields;
+    if (monthCode === undefined) {
+      if (year === undefined && (era === undefined || eraYear === undefined)) {
+        throw new TypeError('`monthCode`, `year`, or `era` and `eraYear` is required');
+      }
+      ({ monthCode, year } = this.adjustCalendarDate({ year, month, monthCode, day, era, eraYear }));
+    }
+
+    let isoYear, isoMonth, isoDay;
+    let closest;
+    // Look backwards starting from 1972 up to 100 years to find a year that
```

Comment:
Is this specified in 402 or TBD?

Reply:
This algorithm must be specified in 402, because it only affects non-ISO calendars. For ISO the reference year is always 1972. (And that *should* be speced in 262!)

That said, I picked 1972 as a starting point because that's the date that the ISO calendar uses, which means that all non-leap-day/non-leap-month dates will use 1972 in all calendars, and Gregorian-based calendars (`gregory`, `japanese`, `roc`, and `buddhist`) will use 1972 for all dates because they share the same leap years as Gregorian.  At least I *think* aligning is an advantage?  A different view could be to deliberately choose a different base year for each calendar to prevent developers from assuming that the reference year will always be 1972 across calendars. I don't have a strong opinion here either way. Leaving as 1972 base for all calendars is easiest so I'd have a slight preference for as-is.

Reply:
Ah, another question: is ISO 1872-1972 sufficient to ensure every ICU calendar will reach all known months and days?

Overall, I don't think this should block #1245 at all, just like zero-padded month codes shouldn't.
@ryzokuken is there an issue for 402 you're aware of that's tracking this? The polyfill will need to be updated when it is implemented. If there's no issue, let's create one somewhere but make sure it's clear it's not a temporal-spec-evolution matter.

Reply:
> Ah, another question: is ISO 1872-1972 sufficient to ensure every ICU calendar will reach all known months and days?

AFAICT, yes. The longest cycle I know about is Chinese's 60 year cycle. @Manishearth, is this correct?

Reply:
> If there's no issue, let's create one somewhere but make sure it's clear it's not a temporal-spec-evolution matter.

I'm nearly sure there's no issue for this in 402.  This is just one of likely many 402 changes which will likely have to be made to align with this PR. The eraYear/era checks would be another, right?  Asserting no skipped days is presumably another. I'm sure I'm forgetting more. Should there be one umbrella issue or a bunch of small ones?

Reply:
Sounds better tracked in an umbrella issue for calendars. We can punt it for now / not worry about it here.

Reply:
OK good. Where should the issue live? In 402 repo or here? @sffc @ryzokuken do you have an opinion?  Here's a starting point: 
* Specify the reference year canonicalization algorithm for monthDayFromFields. Suggestion: start at 1972 and iterate backwards up to 100 years until a year is found where that calendar date exists. If none is found between 1873-1972 then throw a RangeError.
* Specify reference day canonicalization for yearMonthFromFields which is trivial: just always default to `1`
* Specify behavior when era is present without eraYear or vice-versa in xxxFromFields inputs. Suggestion: this should throw.
* Specify behavior when era/eraYear is provided to xxxFromFields when a calendar doesn't use eras. Suggestion: these fields should be ignored (to match ISO)
* Specify that no built-in calendar is allowed to skip any days, like the Julian->Gregorian switchover. (Or if 1+ of them does skip days, then specify that skipping days is OK.)
* Specify behavior of `year` (signed integer, no gaps) and `month` (continuous positive integer values, no gaps) for all built-in calendars

I'm sure I'm forgetting more.

Reply:
(deferring to sffc or ryzokuken; will keep thread open for visibility, but no need to resolve before merging)